test/xbgpu/test_engine.py

```diff
@@ -947,87 +1006,75 @@ def heap_factory(batch_index: int) -> list[spead2.send.HeapReference]:
                 mock_send_stream,
                 corrprod_outputs=corrprod_outputs,
                 beam_outputs=beam_outputs,
-                batch_indices=range(batch_start_index, batch_end_index),
+                batch_indices=test_batch_indices,
                 heap_factory=heap_factory,
                 timestamp_step=timestamp_step,
                 n_ants=n_ants,
                 n_channels_per_substream=n_channels_per_substream,
                 frequency=frequency,
                 n_spectra_per_heap=n_spectra_per_heap,
+                missing_antennas=missing_antennas,
             )
-            last_timestamp = batch_end_index * timestamp_step
-
-        incomplete_accums_counters = []
-        for i, corrprod_output in enumerate(corrprod_outputs):
-            # Or assert if incomplete_accs_total == incomplete_accums_counter * len(xbengine._pipelines)
-            incomplete_accums_counter = 0
-            base_batch_index = batch_start_index
-            for j, corrprod_result in enumerate(corrprod_results[i]):
-                # The first heap is an incomplete accumulation containing a
-                # single batch, we need to make sure that this is taken into
-                # account by the verification function.
-                if j == 0:
-                    # This is to handle the first accumulation processed. The value
-                    # checked here is simply the first in the range.
-                    # - Even though :func:`generate_expected_corrprods` returns a
-                    #   zeroed array for an incomplete accumulation, we still need
-                    #   to maintain programmatic sense in the values generated here.
-                    num_batches_in_current_accumulation = HEAPS_PER_FENGINE_PER_CHUNK
-                    incomplete_accums_counter += 1
-                else:
-                    num_batches_in_current_accumulation = corrprod_output.heap_accumulation_threshold
-                    if missing_antenna is not None:
-                        incomplete_accums_counter += 1
-
-                expected_output = generate_expected_corrprods(
-                    base_batch_index,
-                    num_batches_in_current_accumulation,
-                    corrprod_output.heap_accumulation_threshold,
-                    n_channels_per_substream,
-                    n_ants,
-                    n_spectra_per_heap,
-                    missing_antenna,
-                )
-                base_batch_index += num_batches_in_current_accumulation
-                np.testing.assert_equal(expected_output, corrprod_result)
-            incomplete_accums_counters.append(incomplete_accums_counter)
+            last_timestamp = batch_end_index2 * timestamp_step
+
+        # The first warnings in the list should be from each of the XPipelines,
+        # the rest are warnings related to timestamps between output heaps.
```

Comment:
Can you be sure that they come in this order?

If not, I think caplog has facilities to ignore particular warnings for the duration of the test.

Reply:
Did some digging this morning and found this ~recent PR on pytest that adds the ability to filter within the unit test
- https://github.com/pytest-dev/pytest/pull/11625/files

I see it's part of a newer version of pytest, which we might not necessarily want to upgrade to yet.

I'll use a plain old `filter` for now, maybe create a technical-debt task for updating version and usage.

Reply:
I'd say try updating pytest to the latest and check what happens, but be sure to also test the qualification tests. I vaguely recall that pytest-asyncio 0.23 has some updates to make it compatible with pytest 8, and we're stuck on pytest-asyncio 0.21 until they fix an issue with how scoped fixtures work that breaks the qualification tests.

Reply:
I kicked off a qualification run (on the site Jenkins) around 13:15 today with katgpucbf/upgrade-pytest.

Will likely have to revisit this tomorrow morning. I've done a normal `filter` for now.

Reply:
The qualification run seems to have passed with no complaints/failures. I'll send you the build URL via Mattermost.

Guess I have a bit more work to do before I can re-request a review...